app/data_providers/dispatch.py

```diff
@@ -0,0 +1,73 @@
+import itertools
+import random
+import uuid
+from asyncio import gather
+
+from app.data_providers.curation_api_client import CurationAPIClient
+from app.data_providers.metrics_client import MetricsClient
+from app.data_providers.slate_provider import SlateProvider
+from app.models.corpus_item_model import CorpusItemModel
+from app.models.metrics.firefox_new_tab_metrics_factory import FirefoxNewTabMetricsFactory
+from app.models.ranked_corpus_items_instance import RankedCorpusItemsInstance
+from app.rankers.algorithms import firefox_thompson_sampling_1day
+
+
+class Dispatch:
+    """
+    This class is responsible for accepting:
+
+     a dependency to get items to rank (api_client),
+     a strategy for ranking them (slate_provider),
+     and the data to execute ranking (metrics_client),
+
+     and then using those three things to shape the list of items to the order and size we serve to a client.
+
+     If there are NO rankers, it sends the original list of items, with no reshaping, to the client.
+    """
+    def __init__(
+            self,
+            api_client: CurationAPIClient,
+            slate_provider: SlateProvider,
+            metrics_client: MetricsClient
+    ):
+        self.api_client = api_client
+        self.slate_provider = slate_provider
+        self.metrics_client = metrics_client
+
+    async def get_ranked_corpus_slate(self, slate_id, start_date=None, user_id=None) -> RankedCorpusItemsInstance:
+        corpus_slate_schema = self.slate_provider.get(slate_id)
+
+        # Choose an Experiment
+        # TODO: Implement weighting
+        experiment = random.choice(corpus_slate_schema.experiments)
+
+        # Fetch Corporeal Candidates
+        aggregate_corpus_response = await gather(*(
+            self.api_client.get_ranked_corpus_slate(corpus_id, start_date, user_id)
+            for corpus_id in experiment.eligible_corpora)
+        )
+
+        flattened_unranked_items = list(itertools.chain(*aggregate_corpus_response))
+
+        # If no rankers, return the same list in the same order that the corpus api handed us
+        ranked_items = flattened_unranked_items
+
+        # I do not check for duplicate rankers here. I know the old flow does.
+        # This error is not catastrophic (at most, it reorders items a few times)
+        # It's also not likely (an eng AND their reviewer would have to miss two identical lines next to each other)
+        # And it's not insidious (would be easy to find and verify if we suspected it was happening)
+        # So it's not worth complicating the code to check for IMO
+        for ranker in experiment.rankers:
+            ranker_kwargs = await self.metrics_client.get_engagement_metrics(ranked_items, ranker)
```

Comment:
I was going to put a comment that's similar Jesh's first one, if I understand it correctly. It seems more logical to me the rankers would define their data requirements, instead of metrics clients to know what data to load for a particular ranker. This might look like a ranker using dependency injection to get an object that provides engagement metrics.

Hadn't thought about the second point yet, but that's a good point too. Should be fairly easy to gather everything first, and then map it onto the right ranker.

I don't think we have to do either of these things in this PR. Because there's just one ranker and one type of metrics, it's easy to see how they map, and there wouldn't be a performance benefit to using `gather` yet.

Reply:
So, up until now, the flow we have been using—the one @mmiermans updated a week ago to include the firefox metrics—does this conditionally the same way the metrics client currently does.

I would love to change it. That said, one of the primary goals I had in this PR, based on anxieties expressed by this team, was to _not change_ how rankers worked. That's why I chose _this_ design decision rather than the one suggested.

I confess it's frustrating to have one choice modeled and requested and then another choice suggested after I finish up. I am tabling this idea for a future change set, and we will establish team buy-in before making it.

Reply:
The second point I agree with, but I agree with @mmiermans that it is an implementation detail for the theoretical case we do not have yet in which a ranker requires multiple sets of metrics.